src/main/java/com/google/devtools/build/lib/syntax/MethodLibrary.java

```diff
@@ -170,17 +188,81 @@ private static boolean hasElementWithBooleanValue(
             type = Object.class,
             doc = "This collection.",
             // TODO(cparsons): This parameter should be positional-only.
-            legacyNamed = true)
+            legacyNamed = true),
+        @Param(
+            name = "key",
+            doc = "An optional function applied to each element before comparison.",
+            named = true,
+            defaultValue = "None",
+            positional = false,
+            noneable = true),
+        @Param(
+            name = "reverse",
+            type = Boolean.class,
+            doc = "Return results in descending order.",
+            named = true,
+            defaultValue = "False",
+            positional = false,
+            noneable = true)
       },
       useLocation = true,
       useEnvironment = true)
-  public MutableList<?> sorted(Object self, Location loc, Environment env) throws EvalException {
-    try {
-      return MutableList.copyOf(
-          env, EvalUtils.SKYLARK_COMPARATOR.sortedCopy(EvalUtils.toCollection(self, loc, env)));
-    } catch (EvalUtils.ComparisonException e) {
-      throw new EvalException(loc, e);
+  public MutableList<?> sorted(Object self, Object key, Boolean reverse,
+      Location loc, Environment env)
+      throws EvalException, InterruptedException {
+
+    ArrayList list = new ArrayList(EvalUtils.toCollection(self, loc, env));
+    if (key == Runtime.NONE) {
+      try {
+        Collections.sort(list, EvalUtils.SKYLARK_COMPARATOR);
+      } catch (EvalUtils.ComparisonException e) {
+        throw new EvalException(loc, e);
+      }
+    } else {
+      checkValidKeyFunc(key, loc);
+
+      final Object KEY = key;
+      final Location LOC = loc;
+      final Environment ENV = env;
+      final FuncallExpression AST = new FuncallExpression(Identifier.of(""), ImmutableList.of());
+
+      class KeyComparator implements Comparator<Object> {
+        private Exception e;
```

Comment:
You don't need 'private' here; this class is tailor-made for one function, so there's no need to hide anything.